src/connection/services/Vcard.ts

```diff
@@ -17,11 +18,129 @@ export default class Vcard extends AbstractService {
       return this.sendIQ(iq).then(this.parseVcard);
    }
 
+   public setAvatar(jid: IJID, avatar: string, mimetype: string)
+   {
+      let getVcardIQ = $iq({
+         type: 'get',
+         to: jid.bare
+      }).c('vCard', {
+         xmlns: NS.get('VCARD')
+      });
+
+      //first get the actual vcard to merge image into
+      return this.sendIQ(getVcardIQ).then((stanza) => {
+          if ($(stanza).attr('type')==='result')
+          {
+              let vcard = $(stanza).find('vCard');
+
+              vcard = $($.parseXML(vcard.get(0).outerHTML)).find('>vCard');
+
+              if (!vcard.length) {
+                 //XML BECAUSE OF CASE SENSIVITY
+                 vcard = $($.parseXML('<vCard '+NS.get('VCARD')+'/>').documentElement);
+              }
+
+              let photo = vcard.children('PHOTO');
+              if (avatar!=null&&mimetype!=null)
+              {
+                  let type;
+                  let binval;
+                  if (photo.length===0)
+                  {//XML BECAUSE OF CASE SENSIVITY
+                      photo = $($.parseXML('<PHOTO/>').documentElement);
+                      type = $($.parseXML('<TYPE/>').documentElement);
+                      binval = $($.parseXML('<BINVAL/>').documentElement);
+                      photo.append(type);
+                      photo.append(binval);
+                      vcard.append(photo);
+                  }
+                  else
+                  {
+                      type = photo.children('TYPE');
+                      binval = photo.children('BINVAL');
+                  }
+                  type.text(mimetype);
+                  binval.text(avatar);
+              }
+              else
+              {
+                  if (photo.length>0)
+                  {
+                      vcard.children('PHOTO').remove();
+                  }
+              }
+
+              return this.sendVCard(jid,avatar, vcard, getVcardIQ);
+          }
+          else
+          {
+              return false;
+          }
+      });
+   }
+
+   private sendVCard(jid : IJID, avatar: string, vcard, getVcardIQ): Promise<any>
+   {
+       let sendVcardIQ = $iq({
+                 type: 'set',
+                 to: jid.bare
+              }).cnode(vcard[0]);
+
+        //second send the new vCard with avatar to server
+        return this.sendIQ(sendVcardIQ).then((result)=>{
+          if ($(result).attr('type')==='result')
```

Comment:
Resolved promise should always from type result

Reply:
ok